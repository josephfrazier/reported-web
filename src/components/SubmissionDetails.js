import React from 'react';
import PropTypes from 'prop-types';
import Loadable from 'react-loadable';
import axios from 'axios';
import humanizeString from 'humanize-string';

class SubmissionDetails extends React.Component {
  constructor(props) {
    super(props);

    const { isDetailsOpen } = props;

    this.state = {
      isDetailsOpen,
    };
  }

  render() {
    const {
      reqnumber,
      medallionNo,
      typeofcomplaint,
      timeofreport,

      photoData0,
      photoData1,
      photoData2,

      videoData0,
      videoData1,
      videoData2,
    } = this.props.submission;

    const humanTimeString = new Date(timeofreport).toLocaleString();

    const ImagesAndVideos = () => {
      const images = [photoData0, photoData1, photoData2]
        .filter(item => !!item)
        .map((photoData, i) => {
          const { url } = photoData;
          return (
            <a key={url} href={url} target="_blank" rel="noopener noreferrer">
              <img src={url} alt={`#${i}`} />
            </a>
          );
        });

      const videos = [videoData0, videoData1, videoData2]
        .filter(item => !!item)
        .map((videoData, i) => {
          const url = videoData;
          return (
            <a key={url} href={url} target="_blank" rel="noopener noreferrer">
              {/* eslint-disable-next-line jsx-a11y/media-has-caption */}
              <video src={url} alt={`#${i}`} />
            </a>
          );
        });

      return (
        <React.Fragment>
          {images}
          {videos}
        </React.Fragment>
      );
    };

    const LoadableServiceRequestStatus = Loadable({
      loader: () =>
        axios.get(`/srlookup/${reqnumber}`).then(({ data }) => () => {
          const { error, threeOneOneSRLookupResponse } = data;
          if (error) {
            const { errorMessage, errorCode } = error;
            return `${errorMessage} (error code ${errorCode})`;
          }

          const items = Object.entries(threeOneOneSRLookupResponse[0]).map(
            ([key, value]) => (
              <React.Fragment key={key}>
                <dt>{humanizeString(key)}:</dt>
                <dd>
                  {key.endsWith('Date') ? new Date(value).toString() : value}
                </dd>
              </React.Fragment>
            ),
          );
          return <dl>{items}</dl>;
        }),
      loading: () => 'Loading Service Request Status...',
    });

    return (
      <details
        open={this.state.isDetailsOpen}
        onToggle={evt => {
          this.setState({
            isDetailsOpen: evt.target.open,
          });
        }}
      >
        <summary>
          {medallionNo} {typeofcomplaint} on {humanTimeString}
        </summary>

        {this.state.isDetailsOpen && (
          <React.Fragment>
            <ImagesAndVideos />

            {!reqnumber.startsWith('N/A') && (
              <div>
                <LoadableServiceRequestStatus />
              </div>
            )}
          </React.Fragment>
        )}
      </details>
    );
  }
}

SubmissionDetails.propTypes = {
  isDetailsOpen: PropTypes.bool,
  submission: PropTypes.shape({
    reqnumber: PropTypes.string,
    medallionNo: PropTypes.string,
    typeofcomplaint: PropTypes.string,
    timeofreport: PropTypes.string,

    photoData0: PropTypes.object,
    photoData1: PropTypes.object,
    photoData2: PropTypes.object,

    videoData0: PropTypes.string,
    videoData1: PropTypes.string,
    videoData2: PropTypes.string,
  }).isRequired,
};

SubmissionDetails.defaultProps = {
  isDetailsOpen: false,
};

export default SubmissionDetails;
