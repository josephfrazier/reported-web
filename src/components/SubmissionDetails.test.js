/* eslint-env jest */
/* eslint-disable padded-blocks, no-unused-expressions */

import '@babel/polyfill';
import React from 'react';
import renderer from 'react-test-renderer';
import App from './App';
import SubmissionDetails from './SubmissionDetails';

require('timezone-mock').register('US/Eastern');
require('jest-mock-now')();

describe('SubmissionDetails', () => {
  test('renders children correctly', () => {
    const submission = {
      reqnumber: 'reqnumber',
      medallionNo: 'medallionNo',
      typeofcomplaint: 'typeofcomplaint',
      loc1_address: '82 Reade St, New York, NY 10007, USA',
      timeofreport: new Date(Date.now()).toISOString(),
      reportDescription: 'reportDescription',

      photoData0: { url: 'photoData0.url' },
      photoData1: { url: 'photoData1.url' },
      photoData2: { url: 'photoData2.url' },

      videoData0: 'videoData0',
      videoData1: 'videoData1',
      videoData2: 'videoData2',
    };

    const wrapper = renderer
      .create(
        <App context={{ insertCss: () => {}, fetch: () => {}, pathname: '' }}>
          <SubmissionDetails isDetailsOpen submission={submission} />
        </App>,
      )
      .toJSON();

    expect(wrapper).toMatchSnapshot();
  });
});
