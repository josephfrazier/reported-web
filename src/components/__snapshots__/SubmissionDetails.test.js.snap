// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SubmissionDetails renders children correctly 1`] = `
<details
  onToggle={[Function]}
  open={true}
>
  <summary>
    medallionNo
     
    typeofcomplaint
     near
     
    82 Reade St
     on 
    Mockday 2016-11-18T00:00:00.000Z GMT-0500 (MockDate)
    <br />
    TLC Service Request Number: 
    reqnumber
  </summary>
  <p>
    reportDescription
  </p>
  <a
    href="photoData0.url"
    rel="noopener noreferrer"
    target="_blank"
  >
    <img
      alt="#0"
      src="photoData0.url"
    />
  </a>
  <a
    href="photoData1.url"
    rel="noopener noreferrer"
    target="_blank"
  >
    <img
      alt="#1"
      src="photoData1.url"
    />
  </a>
  <a
    href="photoData2.url"
    rel="noopener noreferrer"
    target="_blank"
  >
    <img
      alt="#2"
      src="photoData2.url"
    />
  </a>
  <a
    href="videoData0"
    rel="noopener noreferrer"
    target="_blank"
  >
    <video
      alt="#0"
      src="videoData0"
    />
  </a>
  <a
    href="videoData1"
    rel="noopener noreferrer"
    target="_blank"
  >
    <video
      alt="#1"
      src="videoData1"
    />
  </a>
  <a
    href="videoData2"
    rel="noopener noreferrer"
    target="_blank"
  >
    <video
      alt="#2"
      src="videoData2"
    />
  </a>
  <div>
    Loading Service Request Status...
  </div>
</details>
`;

exports[`SubmissionDetails renders delete button correctly 1`] = `
<details
  onToggle={[Function]}
  open={true}
>
  <summary>
    medallionNo
     
    typeofcomplaint
     near
     
    82 Reade St
     on 
    Mockday 2016-11-18T00:00:00.000Z GMT-0500 (MockDate)
    <br />
    TLC Service Request Number: 
    reqnumber
  </summary>
  <p>
    reportDescription
  </p>
  <a
    href="photoData0.url"
    rel="noopener noreferrer"
    target="_blank"
  >
    <img
      alt="#0"
      src="photoData0.url"
    />
  </a>
  <a
    href="photoData1.url"
    rel="noopener noreferrer"
    target="_blank"
  >
    <img
      alt="#1"
      src="photoData1.url"
    />
  </a>
  <a
    href="photoData2.url"
    rel="noopener noreferrer"
    target="_blank"
  >
    <img
      alt="#2"
      src="photoData2.url"
    />
  </a>
  <a
    href="videoData0"
    rel="noopener noreferrer"
    target="_blank"
  >
    <video
      alt="#0"
      src="videoData0"
    />
  </a>
  <a
    href="videoData1"
    rel="noopener noreferrer"
    target="_blank"
  >
    <video
      alt="#1"
      src="videoData1"
    />
  </a>
  <a
    href="videoData2"
    rel="noopener noreferrer"
    target="_blank"
  >
    <video
      alt="#2"
      src="videoData2"
    />
  </a>
  <button
    onClick={[Function]}
    style={
      Object {
        "background": "white",
        "color": "red",
        "margin": "1px",
      }
    }
    type="button"
  >
    <span
      aria-label="Delete Submission"
      role="img"
    >
      ‚ùå
    </span>
    Delete Submission
  </button>
</details>
`;
